### ISSUE TITLE
Support OIDC for CNAMEs

### ISSUE THREAD

**Issue by hexylena** [galaxyproject/galaxy#8215]
Currently we hardcode the `redirect_uri` parameter in the OIDC configuration. I would like to propose that we allow an alternative configuration that looks like this:

```xml
<?xml version="1.0"?>
<OIDC>
        <provider name="Elixir">
                <client_id></client_id>
                <client_secret></client_secret>
                <redirect_path>/authnz/elixir/callback</redirect_uri>
                <prompt>consent</prompt>
        </provider>
</OIDC>
```

Where `redirect_path` is used instead of `redirect_uri`. I would then like Galaxy to construct the domain + protocol portion of the URI based on the domain + protocol of the request.

I'm proposing this because of the following scenario: UseGalaxy.eu has ~10 and increasing CNAMEs for the same server, all responded to by the same process, sometimes with different content on subdomains (by proxying + the built in subdomain/cname features of galaxy like showing different toolboxes.) Our OIDC client is configured such that login from any of these subdomains will work, we just need to respond with a different redirect_uri

This would let us have login working properly across subdomains, and logins on subdomains would not be redirected to the parent domain.

cc @bgruening 

Would you be open to this or maybe have time for this change @VJalili?

**Comment by VJalili** [comment]
I don't like the `redirect_uri` either, since it should be possible to reconstruct it automatically. I have even tried removing it from the config (so you would not need to specify it) and construct it on-the-fly (see https://github.com/galaxyproject/galaxy/pull/7620 that leverages `trans.request.host_url`). However, it seems there is no solid way of determining the base URI at which your Galaxy instance is running. 

Therefore, I would like to add a parameter (e.g., `base_uri`) to [`oidc_config`](https://github.com/galaxyproject/galaxy/blob/dev/config/oidc_config.xml.sample) (where a set of setters used for all the backends are configured) and remove `redirect_uri` from [`oidc_backends_config`](https://github.com/galaxyproject/galaxy/blob/dev/config/oidc_backends_config.xml.sample), hence to enable generating the callback automatically. (I should be able to PR that sometime in July, if no objection to the approach.) 

With that you will have something like this: 

- an example of `oidc_backends_config.xml` where `redirect_uri` is removed for all the backends:
    ```xml
    <?xml version="1.0"?>
    <OIDC>
        <provider name="Elixir">
            <client_id></client_id>
            <client_secret></client_secret>
            <prompt>consent</prompt>
        </provider>
    </OIDC>
    ```

- an example of `oidc_config.xml` where `BASE_URI` is added:
    ```xml
    <?xml version="1.0"?>
    <OIDC>
        <Setter Property="VERIFY_SSL" Value="False" Type="bool"/>
        <Setter Property="REQUESTS_TIMEOUT" Value="3600" Type="float"/>
        <Setter Property="ID_TOKEN_MAX_AGE" Value="3600" Type="float"/>
        <Setter Property="BASE_URI" Value="localhost:8080" Type="str"/>
    </OIDC>
    ```

Or if you have multiple values for `BASE_URI` probably we can allow the value of the setter to be a list, and pick from the list based on the base URI value in the callback request.

Would that make sense? or any suggestion/alternative?

**Comment by hexylena** [comment]
Interesting that you've already tried this! I had not seen that PR previously. I guess I can understand the rationale for closing that, but surely it should be the admin's responsibility to ensure the proper request headers are sent? I would argue that we should just do this, and if it is broken for those admin, then it's their responsibility.

I  your post above you wrote `trans.request.host_url`, Is it then different with URL? Or is that just protocol + fqdn properly concatenated? The specific post you linked seems to be about detecting if it was proxied more than just responding to the request as it was sent. If `request.host` is reading the `host` header, and if this is set incorrectly, then the admin should just change their proxy configuration.

E.g. I hot patched one of our servers to just print `request.host`:

![image](https://user-images.githubusercontent.com/458683/60176390-6a33e880-9816-11e9-9c43-916decab593b.png)
![image](https://user-images.githubusercontent.com/458683/60176406-7455e700-9816-11e9-8a24-5803b2742f21.png)

Since it's http proxied and not uwsgi proxied, the only proxying change required was:

```
proxy_set_header Host $host;
```

I tested uwsgi sockets as well, and there don't even need to set the header, it's done correctly by default. I guess in the end my point is: the PR you originally opened *could* work for some of us, and for those it couldn't, they would just need to 1) update their proxying settings. and 2) actually test the OIDC that they roll out, rather than just waiting on user error reports.

**Comment by VJalili** [comment]
Thanks for confirming that, I am happy to reopen the PR https://github.com/galaxyproject/galaxy/pull/7620 (or alternatively create a similar one). 

What do you think about the second method I posted an example of? this can potentially work for all scenarios, and one config is used for all the authnz backends.

**Comment by hexylena** [comment]
Picking from the list is an interesting idea, but it would add an extra barrier for us to deployment? Currently our process for launching a new subdomain is one step: setup a DNS entry. With that we would have to edit that list every time and restart the server.

Is there a scenario you had in mind where using the request host wouldn't function correctly?

**Comment by VJalili** [comment]
> Is there a scenario you had in mind where using the request host wouldn't function correctly?

No, I did not see any issues myself, though I found some online discussions where it seemed the consensus is that, it is not guaranteed. 

@natefoo @martenson any thoughts?

**Comment by VJalili** [comment]
@erasche when you define multiple DNS entries, are you defining those entries with the IdP? 

For instance, if you define: 
```
usegalaxy.eu/authnz/elixir/callback
training.usegalaxy.eu/authnz/elixir/callback
```
both of these endpoints need to be defined with the IdP.

**Comment by hexylena** [comment]
@VJalili yes, we have all of those entrypoints registered. (Ok, you're right, there was a second step to setting it up, had not thought about that one.)

Our client configuration looks nice with our IdP :)

![image](https://user-images.githubusercontent.com/458683/60202057-1b07ab00-984a-11e9-82ca-26a12b0706a5.png)

**Comment by VJalili** [comment]
> Our client configuration looks nice with our IdP :)

indeed :)
